apps/site/pages/en/learn/test-runner/mocking.md

```diff
@@ -0,0 +1,270 @@
+
+Node.js provides many ways to mock various pieces of code.
+
+But first, there are several types of tests:
+
+| type             | description                               | example                                                                                        | mock candidates                          |
+| :--------------- | :---------------------------------------- | :--------------------------------------------------------------------------------------------- | :--------------------------------------- |
+| unit             | the smallest bit of code you can isolate  | `const sum = (a, b) => a + b`                                                                  | own code, external code, external system |
+| component        | a unit + dependencies                     | `const arithmetic = (op = sum, a, b) => ops[op](a, b)`                                         | external code, external system           |
+| integration      | components fitting together               | -                                                                                              | external code, external system           |
+| end-to-end (e2e) | app + external data stores, delivery, etc | A fake user (ex a Playwright agent) literally using an app connected to real external systems. | none (do not mock)                       |
+
+There are varying schools of thought about when you should and should not mock, the broadstrokes of which are laid out below.
+
+## When and not to mock
+
+There are 3 main mock candidates:
+
+- own code
+- external code
+- external system
+
+### Own code
+
+This is what you/your project control.
+
+```mjs displayName="your-project/main.mjs"
+import foo from './foo.mjs';
+
+export function main() {
+  const f = foo();
+}
+```
+
+Here, `foo` is an "own code" dependency of `main`.
+
+#### Why
+
+For a true unit test of `main`, `foo` should be mocked: you're testing that `main` works, not that `main` + `foo` work (that's a different test).
+
+#### Why not
+
+Mocking `foo` can be more trouble than worth, especially when `foo` is simple, well-tested, and rarely updated.
+
+Others argue that not mocking `foo` is better because it's more authentic and increases coverage of `foo` (because `main`'s tests will also verify `foo`). This can, however, create noise: when `foo` breaks, a bunch of other tests will also break, so tracking down the problem is more tedious: if only the 1 test for the item ultimately responsible for the issue is failing, that's very easy to spot; whereas 100 tests failing creates a needle-in-a-haystack to find the real problem.
```

Comment:
```suggestion
Mocking `foo` can be more trouble than worth, especially when `foo` is simple, well-tested, and rarely updated.

Not mocking `foo` can be better because it's more authentic and increases coverage of `foo` (because `main`'s tests will also verify `foo`). This can, however, create noise: when `foo` breaks, a bunch of other tests will also break, so tracking down the problem is more tedious: if only the 1 test for the item ultimately responsible for the issue is failing, that's very easy to spot; whereas 100 tests failing creates a needle-in-a-haystack to find the real problem.
```